README.md
[![Build Status](http://komanda.io:8080/api/badge/github.com/mephux/envdb/status.svg?branch=master)](http://komanda.io:8080/github.com/mephux/envdb)

# Envdb - Environment Database

* NOTE: envdb is still beta software.

<img style="float:left;" height="244px" src="https://raw.githubusercontent.com/mephux/envdb/master/data/envdb.gif">

Envdb turns your production, dev, cloud, etc environments into a database 
cluster you can search using [osquery](https://github.com/facebook/osquery) as the foundation.

Envdb allows you to register each computer, server or asset as a node in a cluster. Once a new
node is connected it becomes available for search from the Envdb ui. 
Envdb was built using golang so the whole application, node client and server comes as one single binary.
This makes it really easy to deploy and get working in seconds.

Video Intro: [https://youtu.be/ydYr7Ykwzy8](https://youtu.be/ydYr7Ykwzy8)

## How it works.

Envdb wraps the osquery process with a node agent (node as in cluster node) that can communicate back to a central location.
When that node gets a new query, it's executed and then sent back to the tcp server for rendering. Once the
request is processed it's then sent to any avaliable web clients using websockets.

Envdb has an embedded sqlite database for node storage and saved searches.

ui --websockets--> server --tcp--> node client.

## Download

Pre-built versions (deb/tar.gz) of envdb are avaliable for linux 386/amd64. 
[linux downloads](https://github.com/mephux/envdb/releases)

Building on macosx is easy tho, checkout the section below.

## Building

  Make sure you have Go installed. I used Go version 1.4.1.

#### deps

  * `go get github.com/jteeuwen/go-bindata/...`
  * `go get github.com/elazarl/go-bindata-assetfs`
  * `go get github.com/tools/godep`

#### make

  * `git clone https://github.com/mephux/envdb.git`
  * `cd envdb`
  * `make`

## Usage

  * **NOTICE**: The default username and password for the UI.
    * username: **admin@envdb.io**
    * password: **envdb**

```
usage: envdb [<flags>] <command> [<flags>] [<args> ...]

The Environment Database - Ask your environment questions

Flags:
  --help       Show help.
  --debug      Enable debug logging.
  --dev        Enable dev mode. (read assets from disk and
               enable debug output)
  -q, --quiet  Remove all output logging.
  --version    Show application version.

Commands:
  help [<command>]
    Show help for a command.

  server [<flags>]
    Start the tcp server for node connections.

  node --server=127.0.0.1 [<flags>] <node-name>
    Register a new node.

  users [<flags>]
    User Management (Default lists all users).
```

#### Server

```
$ envdb help server

  usage: envdb [<flags>] server [<flags>] [<command>]

  Start the tcp server for node connections.

  Flags:
    -p, --port=3636  Port for the server to listen on.
    -P, --http-port=8080  
                     Port for the web server to listen on.

  Args:
    [<command>]  Daemon command. (start,status,stop)
```

    Running the server without a `command` (start, stop or status) will run the server in the foreground.

    * Note: By default this will start the tcp server on port 3636 and the web server on port 8080.

#### Node Client

```
$ envdb help node

  usage: envdb [<flags>] node --server=127.0.0.1 [<flags>] <node-name>

  Register a new node.

  Flags:
    -s, --server=127.0.0.1  
                     Address for server to connect to.
    -p, --port=PORT  Port to use for connection.

  Args:
    <node-name>  A name used to uniquely identify this node.
```

    `sudo envdb node --server <ip to server> SomeBoxName`

  * That's it - it's really that simple.

#### User Management

  * list users
    `envdb users`

  * add a new user
    `envdb users --add`

  * remove a new user
    `envdb users --remove <email>`

## More UI

<img style="float:left;" height="300px" src="https://raw.githubusercontent.com/mephux/envdb/master/data/envdb-1.png">

<img style="float:left;" height="300px" src="https://raw.githubusercontent.com/mephux/envdb/master/data/envdb-2.png">

<img style="float:left;" height="300px" src="https://raw.githubusercontent.com/mephux/envdb/master/data/envdb-3.png">

## Self-Promotion

Like envdb? Follow the repository on
[GitHub](https://github.com/mephux/envdb) and if
you would like to stalk me, follow [mephux](http://dweb.io/) on
[Twitter](http://twitter.com/mephux) and
[GitHub](https://github.com/mephux).

## TODO

  * add in memeory pagination for results.
  * Node/Server auth, verification and validation.
  * Code cleanup (will continue forever).


CHANGELOG.md
# Changelog

## 0.4.1 2015-05-13

  + Add loading indicator for system information 
    request
  + bug fixes
  + store envdb version for nodes
  + Allow nodes list to scroll if needed.
  + Update MakeFile version parsing logic.
  + Revert removing debug symbols from build

## 0.4.0 2015-04-18

  + Add system information window for node.
    Right click node and select `system information`.
  + bug fixes

## 0.3.1 2015-04-18

  + Fix table sort - respect query order by first.
  + perf limit added to query response. Limit by 10k.
  + better query debug output for node scope queries.
  + minor UI fixes.
  + rename osquery.go to query.go

## 0.3.0 - 2015-04-16

  + fix bug with node template not showing os icon
  + when clicking a table in a node view - add select query to editor
  + Truncate node metadata in node listing.
  + Editor comments, new lines and pop last valid for current search.
  + Way better error messages for query syntax issues.
  + Add a few new default queries:
    * Third-party kernel extensions (OS X)
    * Startup items (OS X)
    * Interface information
    * Shell history
    * All users with groups
  + minor bug fixes

  - socket verbose output for errors returned from handlers

## 0.2.4 - 2015-04-16

  + better debug output for node request/time of query
  + daemon support for server
  + minor bug fixes

## 0.2.3 - 2015-04-13

  + os icon for node button
  + better query loading indicator.
  + readme updates


envdb.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"runtime"
	"strings"

	"github.com/howeyc/gopass"
	"gopkg.in/alecthomas/kingpin.v1"
)

const (
	// Name application name
	Name = "envdb"

	// Version application version number
	Version = "0.4.1"

	// DefaultServerPort the default tcp server port
	DefaultServerPort = 3636

	// DefaultWebServerPort the default web server port
	DefaultWebServerPort = 8080
)

var (
	// TimeFormat global time format string
	TimeFormat = "15:04:05"

	app   = kingpin.New(Name, "The Environment Database - Ask your environment questions")
	debug = app.Flag("debug", "Enable debug logging.").Short('v').Bool()
	dev   = app.Flag("dev", "Enable dev mode.").Bool()
	quiet = app.Flag("quiet", "Remove all output logging.").Short('q').Bool()

	server        = app.Command("server", "Start the tcp server for node connections.")
	serverCommand = server.Arg("command", "Daemon command. (start,status,stop)").String()
	// serverConfig = server.Flag("config", "Server configuration file.").File()
	serverPort = server.Flag("port", "Port for the server to listen on.").
			Short('p').PlaceHolder(fmt.Sprintf("%d", DefaultServerPort)).Int()

	serverWebPort = server.Flag("http-port", "Port for the web server to listen on.").
			Short('P').PlaceHolder(fmt.Sprintf("%d", DefaultWebServerPort)).Int()

	node = app.Command("node", "Register a new node.")
	// clientConfig = client.Flag("config", "Client configuration file.").File()
	nodeName = node.Arg("node-name", "A name used to uniquely identify this node.").Required().String()

	nodeServer = node.Flag("server", "Address for server to connect to.").
			Short('s').PlaceHolder("127.0.0.1").Required().String()

	nodePort = node.Flag("port", "Port to use for connection.").Short('p').Int()

	users      = app.Command("users", "User Management (Default lists all users).")
	addUser    = users.Flag("add", "Add a new user.").Short('a').Bool()
	removeUser = users.Flag("remove", "Remove user by email.").Short('r').PlaceHolder("email").String()

	// Log Global logger
	Log *Logger

	// DevMode Development mode switch. If true
	// debug logging and serving assets from disk
	// is enabled.
	DevMode bool

	// TestMode
	TestMode bool
)

func initLogger() {
	Log = NewLogger()

	Log.Prefix = Name

	if *debug {
		Log.SetLevel(DebugLevel)
	} else {
		Log.SetLevel(InfoLevel)
	}

	if *dev {
		DevMode = true
		Log.SetLevel(DebugLevel)
		Log.Info("DEBUG MODE ENABLED.")
	} else {
		DevMode = false
	}

	if *quiet {
		Log.SetLevel(FatalLevel)
	}

}

func main() {
	runtime.GOMAXPROCS(runtime.NumCPU())

	app.Version(Version)
	args, err := app.Parse(os.Args[1:])

	initLogger()

	switch kingpin.MustParse(args, err) {

	case users.FullCommand():
		serverSetup(false)

		if *addUser {
			addDBUser()
			return
		}

		if len(*removeUser) > 0 {
			if user, err := FindUserByEmail(*removeUser); err != nil {
				Log.Fatal(err)
			} else {
				if err := user.Delete(); err != nil {
					Log.Fatal(err)
				}
			}

			Log.Info("User removed successfully.")
			return
		}

		users, err := FindAllUsers()

		if err != nil {
			Log.Fatal(err)
		}

		fmt.Println("Listing Users: ")
		for _, user := range users {
			fmt.Printf("  * %s (%s)\n", user.Name, user.Email)
		}

	case server.FullCommand():

		serverSetup(true)

	case node.FullCommand():

		output, err := exec.Command("whoami").Output()

		if err != nil {
			Log.Fatalf("Error: %s", err)
			os.Exit(-1)
		}

		if strings.Trim(string(output), "\n") != "root" {
			Log.Fatal("You must run the node client as root.")
			os.Exit(-1)
		}

		var clntPort = DefaultServerPort

		if *nodePort != 0 {
			clntPort = *nodePort
		}

		var c = Node{
			Name:       *nodeName,
			Host:       *nodeServer,
			Port:       clntPort,
			RetryCount: 50,
		}

		config, err := NewNodeConfig()

		if err != nil {
			Log.Fatal(err)
		}

		c.Config = config

		if err := c.Run(); err != nil {
			Log.Error(err)
		}

	default:
		app.Usage(os.Stdout)
	}

}

func serverSetup(start bool) {
	var svrPort = DefaultServerPort
	var svrWebPort = DefaultWebServerPort

	if *serverPort != 0 {
		svrPort = *serverPort
	}

	if *serverWebPort != 0 {
		svrWebPort = *serverWebPort
	}

	svr, err := NewServer(svrPort)

	if err != nil {
		Log.Fatal(err)
	}

	if !start {
		return
	}

	if len(*serverCommand) <= 0 {
		if err := svr.Run(svrWebPort); err != nil {
			Log.Error(err)
		}
	} else {

		switch *serverCommand {
		case "start":
			svr.Config.Daemon.StartServer(svr, svrWebPort)
			break
		case "stop":
			svr.Config.Daemon.Stop()
			break
		case "status":
			svr.Config.Daemon.Status()
			break
		default:
			{
				Log.Fatalf("Error: Unknown Command %s.", *serverCommand)
			}
		}

	}

}

func ask(reader *bufio.Reader, question string) string {
	fmt.Print(question)

	value, _ := reader.ReadString('\n')
	trim := strings.Trim(value, "\n")

	if len(trim) <= 0 {
		Log.Fatalf("value cannot be blank.")
	}

	return trim
}

func addDBUser() {
	reader := bufio.NewReader(os.Stdin)

	name := ask(reader, "Name: ")

	email := ask(reader, "Email ")

	if !IsEmail(email) {
		Log.Fatalf("%s is not a valid email address.", email)
	}

	fmt.Print("Password: ")
	pass := gopass.GetPasswd()

	fmt.Print("Confirm: ")
	cpass := gopass.GetPasswd()

	if string(pass) != string(cpass) {
		Log.Fatal("Password and confirm do not match.")
	}

	user := &UserDb{
		Name:     name,
		Email:    email,
		Password: string(pass),
	}

	err := CreateUser(user)

	if err != nil {
		Log.Fatal(err)
	}

	Log.Info("User created successfully.")
}
